### ISSUE TITLE
Implement "Amplitude" Relationship Type

### ISSUE THREAD

**Issue by jenncross** [CMU-CREATE-Lab/flutter-app-android#79]
Amplitude= amplitude of oscillation is linearly proportional to the sensor value.

First needs to be added to the firmware and protocol...

**Comment by jenncross** [comment]
(I'll copy this to github also for future reference)

Yeah, amplitude control is a hard thing to specify.... probably the hardest that we have.

I did some brainstorming around it way back when.... trying to think of all the variables needed to define the transform equations that each link really is.

Some of that is here:
https://docs.google.com/document/d/1mLc7D9V8lG76d_B_vKkIjrMtNTPMWEdydZsMeFqq6yE/edit?usp=sharing
 
And some is on the second tab of:
https://docs.google.com/spreadsheets/d/1Pdok42zFDso3IHy9TEELg1uefztyF-O1w1fP6saD47o/edit?usp=sharing


Generally the output side of things needs two basic specification paradigms, either:
A) A center point and a maximum amplitude
B)A upper bound and lower bound (with the max amplitude and center point being infered)

Also there is a advanced setting need for fixing the frequency of oscillation --- faster or slower LED blinking for example.

----

Looks like I used different paradigm for LEDs and servos .... this was have the servo have:
90 - degrees as the default center point and then a selector for the servo upper bound between 0 and 90 degrees (0 would mean that regardless the servo would stay at the center point and 90 degree max amplitude would mean that the servo would be able to hit both 0 and 180 with the max sensor value.

The behavior is best described as a sinusoid, centered on the center point and with the amplitude being set via that upper bound and the minimum amplitude always being zero (some day we may want to be able to set that. A restriction will be:center point - the bound  >= 0 degrees and center point + bound <= 180

Note: It will be important for the software side to be able to set the high and low sensor bounds like in other link types, so that we can adjust sensitivity and invert the behavior (i.e more waving for lower values).

The LED is easiest to manage with the other paradigm where we assume that we can set the high and low boundaries of the color (default off and white?) and then from there we calculate the center point as the average RGB values between the two. (So that the user could blink between yellow and pink or whatever is pleasing) (It is hard to go the other way from a center point and a amplitude to back out what the "amplitude" is and what the minimum color bound would be.)

I suppose that the two point paradigm will make more sense for the buzzer..... where we set a minimum pitch and maximum pitch, and infer the max amplitude and center point from there? I would suggest that end points could be two octives apart and centering on 440 seems nice so the minimum and maximum pitch would be 220 and 880 hz respectively. That does leave some interesting math though involving the exponential behavior of frequencies and how equal spacing between hz does not equate to equal musical spacing between notes heard --- which is probably a thing worth thinking about how to handle.

It might be worth making the servo behavior and the buzzer volume follow the same paradigm as the LEDs and the buzzer pitch -- to simplify the explanation to teachers and students.


... sorry for the extended musings, I hope you find some of those ideas helpful.

Jenn
